services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/DefaultFileDownload.java

```diff
@@ -28,18 +36,51 @@
 
     private final CompletableFuture<CompletedFileDownload> completionFuture;
     private final TransferProgress progress;
+    private final DownloadFileMonitor monitor;
+    private volatile boolean paused;
 
     DefaultFileDownload(CompletableFuture<CompletedFileDownload> completionFuture,
-                        TransferProgress progress) {
+                        TransferProgress progress,
+                        DownloadFileMonitor monitor) {
         this.completionFuture = completionFuture;
         this.progress = progress;
+        this.monitor = monitor;
     }
 
     @Override
     public TransferProgress progress() {
         return progress;
     }
 
+    @Override
+    public PersistableFileDownload pause() {
+        validatePauseStatus();
+        paused = true;
+        completionFuture.cancel(false);
+        GetObjectResponse getObjectResponse = monitor.initialResponse().get();
+        long bytesTransferred = progress.snapshot().bytesTransferred();
```

Comment:
Yeah, cancelling immediately is not strictly guaranteed, so I think it's possible. It's probably fine to have this discrepancy between the bytes transferred captured here and the bytes actually written in the file since the bytes will be overridden upon resume.

Reply:
Would we ever want to consider inferring the `bytesTransferred` based on the existing, partial file size upon resuming?